src/SlimBootstrap/ResponseOutputWriter/Csv.php

```diff
@@ -0,0 +1,305 @@
+     *
+     * @var string
+     */
+    private $_linebreak = "\r\n";
+
+    /**
+     * CSV NULL
+     *
+     * @var string
+     */
+    private $_null = 'NULL';
+
+    /**
+     * Keyspace delimiter
+     *      Used if a multidimensional array is used to map values to
+     *          "<parentKey><keyspaceDelimiter><childKey>"
+     *      E.g.
+     *          array("user" => array("name" => "foobar"))
+     *      would become
+     *          "user_name"
+     *
+     *
+     * @var string
+     */
+    private $_keyspaceDelimiter  = '_';
+
+    /**
+     * @param Slim\Http\Request  $request  The Slim request object.
+     * @param Slim\Http\Response $response The Slim response object.
+     * @param Slim\Http\Headers  $headers  The Slim response headers object.
+     * @param String             $shortName
+     */
+    public function __construct(
+        Slim\Http\Request $request,
+        Slim\Http\Response $response,
+        Slim\Http\Headers $headers,
+        $shortName
+    ) {
+        $this->_request  = $request;
+        $this->_response = $response;
+        $this->_headers  = $headers;
+    }
+
+    /**
+     * This function outputs the given $data as valid CSV to the client
+     * and sets the HTTP Response Code to the given $statusCode.
+     *
+     * @param SlimBootstrap\DataObject[] $data       The data to output to
+     *                                                   the client
+     * @param int                            $statusCode The status code to set
+     *                                                   in the response
+     */
+    public function write($data, $statusCode = 200)
+    {
+        $result = array();
+
+        if (true === \is_array($data)) {
+            $this->_normalizeAll($data);
+            $identifiers = null;
+            foreach ($data as $entry) {
+                $this->_buildStructure($entry, $entry->getIdentifiers(), 0, $result);
+            }
+        } else {
+            $this->_buildStructure($this->_normalizeOne($data), $data->getIdentifiers(), 0, $result);
+        }
+
+        $body = $this->_csvEncode($result);
+
+        if (false === $body) {
+            $this->_response->setStatus(500);
+            $this->_response->setBody("Error encoding requested data.");
+            return;
+        }
+
+        $this->_headers->set(
+            'Content-Type',
+            'text/csv; charset=UTF-8'
+        );
+
+        $this->_response->setStatus($statusCode);
+        $this->_response->setBody($body);
+    }
+
+    /**
+     * @param SlimBootstrap\DataObject $data  DataObject
+     * @param array|null $keys
+     * @return array
+     */
+    private function _normalizeOne(SlimBootstrap\DataObject &$data, $keys = null){
+        if(null === $keys){
+            $keys[] = \array_keys($data->getData());
+        }
+        $keys = \array_fill_keys($keys, null);
+
+        if (\count(\array_intersect_key($keys, $data->getData())) !== \count($keys)){
```

Comment:
please add a small comment, which explains why you need to update original data.